File: sources/extractor/internal/scip/testdata/fixtures/javascript-express/src/service.js
// CampaignService is the intermediate hop the walker must follow.
//
// See handlers.js for the JSDoc rationale; the same `@type` pattern
// links `this.repo` to CampaignRepository so the walker can step
// from getById → CampaignRepository.findById.

const { CampaignRepository } = require("./repository");

class CampaignService {
  /** @param {CampaignRepository} repo */
  constructor(repo) {
    /** @type {CampaignRepository} */
    this.repo = repo;
  }

  getById(id) {
    if (!id) {
      throw new Error("id required");
    }
    const value = this.repo.findById(id);
    if (!value) {
      throw new Error("not found");
    }
    return value;
  }

  create(id, value) {
    if (!id || !value) {
      throw new Error("invalid");
    }
    this.repo.save(id, value);
  }
}

module.exports = { CampaignService };

File: internal/extractor/scip/testdata/fixtures/python-fastapi/app/service.py
"""CampaignService is the intermediate hop in the call graph."""

from .repository import CampaignRepository


class CampaignService:
    def __init__(self, repo: CampaignRepository) -> None:
        self._repo = repo

    def get_by_id(self, id: str) -> str:
        if not id:
            raise ValueError("id required")
        value = self._repo.find_by_id(id)
        if value is None:
            raise LookupError("not found")
        return value

    def create(self, id: str, value: str) -> None:
        if not id or not value:
            raise ValueError("invalid")
        self._repo.save(id, value)
